### PR TITLE
Reporting Dashboard v2.1

### PR DESCRIPTION
**Pull request by OptimusRhine**
https://www.pivotaltracker.com/story/show/150828841

* moved reporting stat generation to separate model
  * fixed minor errors in stat generation (corrected some checks, etc)
  * added stats for intervention access (not checking if `interv` in `user.interventions`, but rather checking `interv.display_for_user(user).access`)
  * added renaming of 'Decision Support' to 'Decision Support P3P' (just within the stats, not the actual db entries)
* updated reporting template page
  * corrected and added minor details to existing descriptions
  * moved Intervention Report counts to separate table (as users can have interv reports even if they don't have those intervs in their `user.interventions`)
  * added Intervention Access table
  * fixed date checks for registration/logins
  * no longer showing blank-named rows in tables (mostly relevant for ePROMs interventions)
* added celery task for refreshing reporting stats
  * ~**TO DO:** add entry for task to persistence files~ (see comment below)
* added reporting stats tests

### CODE REVIEW

**Comment by OptimusRhine** [comment]
Adding ScheduledJob to TrueNTH:
https://github.com/uwcirg/TrueNTH-USA-site-config/pull/54

File: portal/models/i18n.py [hunk 1]
```diff
@@ -254,6 +254,6 @@ def merge_po_into_master(po_path, language, fname):
 
 @babel.localeselector
 def get_locale():
-    if current_user() and current_user().locale_code:
+    if session and current_user() and current_user().locale_code:
```
Comment (pbugni) [inline portal/models/i18n.py]: Lets move the check for session into `current_user`

File: portal/models/reporting.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    stats['intervention_access'] = defaultdict(int)
+    stats['intervention_reports'] = defaultdict(int)
+    stats['organizations'] = defaultdict(int)
+    stats['registrations'] = []
+    stats['encounters'] = defaultdict(list)
+
+    interventions = Intervention.query.all()
+
+    for user in User.query.filter_by(active=True):
+        if ROLE.TEST in [r.name for r in user.roles]:
```
Comment (pbugni) [inline portal/models/reporting.py]: There is a `user.has_role(ROLE.TEST)` method for this...

File: portal/models/reporting.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                    stats['patients']['dx-t'] += 1
+                if any((obs.codeable_concept == CC.PCaLocalized
+                        and obs.value_quantity == CC.FALSE_VALUE)
+                       for obs in user.observations):
+                    stats['patients']['meta'] += 1
+
+        for interv in interventions:
+            desc = interv.description
+            if desc == 'Decision Support':
+                desc = 'Decision Support P3P'
```
Comment (pbugni) [inline portal/models/reporting.py]: Hmm - there are multiple `decision support` options.  I'm not sure we should be forcing this to P3P here.  Would it be safer to switch on intervention.name?

File: portal/models/reporting.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            if interv.display_for_user(user).access:
+                stats['intervention_access'][desc] += 1
+            if interv in user.interventions:
+                stats['interventions'][desc] += 1
+            if (any(doc.intervention == interv for doc in user.documents)):
+                stats['intervention_reports'][desc] += 1
+
+        if not user.organizations.all():
```
Comment (pbugni) [inline portal/models/reporting.py]: `user.organizations.count()`

File: portal/models/reporting.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+from .procedure_codes import known_treatment_not_started
+from .role import ROLE
+from .user import User
+
+
+@dogpile_cache.region('hourly')
+def get_reporting_stats():
+    """Cachable interface for expensive reporting data queries
+
+    The following code is only run on a cache miss.
```
Comment (pbugni) [inline portal/models/reporting.py]: I know there's logging in the celery task, but I think it'd be good to note this is running (catch any overlooked cache misses).

**Comment by OptimusRhine** [comment]
@pbugni Updated with your suggested changes. Note that I had to modify your recent change to `AssessmentStatus.__str__`, because when testing with eproms-stg data I was getting some AssessmentStatus objects where `self.qb_data.qb == None`

**Pull request merged**